query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+    /// Note that if this function decides that a pattern cannot match, it will flip
+    /// `is_known_empty`.
+    fn alias_table<'s, 'a>(&mut self, schema: &'s Schema, pattern: &'a Pattern) -> Option<SourceAlias> {
+        self.table_for_places(schema, &pattern.attribute, &pattern.value)
+            .map(|table| SourceAlias(table, (self.aliaser)(table)))
+    }
+
+    fn get_attribute<'s, 'a>(&self, schema: &'s Schema, pattern: &'a Pattern) -> Option<&'s Attribute> {
+        match pattern.attribute {
+            PatternNonValuePlace::Entid(id) =>
+                schema.attribute_for_entid(id),
+            PatternNonValuePlace::Ident(ref kw) =>
+                self.attribute_for_ident(schema, kw),
+            _ =>
+                None,
+        }
+    }
+
+    fn get_value_type<'s, 'a>(&self, schema: &'s Schema, pattern: &'a Pattern) -> Option<ValueType> {
+        self.get_attribute(schema, pattern).map(|x| x.value_type)
+    }
+}
+
+/// Expansions.
+impl ConjoiningClauses {
+
+    /// Take the contents of `bindings` and generate inter-constraints for the appropriate
+    /// columns into `wheres`.
+    /// For example, a bindings map like
+    ///
+    /// ```clojure
+    ///   {?foo [:datoms12.e :datoms13.v :datoms14.e]}
+    /// ```
+    ///
+    /// produces additional constraints:
+    ///
+    /// ```clojure
+    /// [[:= :datoms12.e :datoms13.v] [:= :datoms12.e :datoms14.e]]
+    /// ```
+    pub fn expand_bindings(&mut self) {
+        for cols in self.bindings.values() {
+            if cols.len() > 1 {
+                let ref primary = cols[0];
+                let secondaries = cols.iter().skip(1);
+                for secondary in secondaries {
+                    // TODO: if both primary and secondary are .v, should we make sure
+                    // the type tag columns also match?
+                    // We don't do so in the ClojureScript version.
+                    self.wheres.push(ColumnConstraint::EqualsColumn(primary.clone(), secondary.clone()));
+                }
+            }
+        }
+    }
+
+    /// When a CC has accumulated all patterns, generate value_type_tag entries in `wheres`
+    /// to refine value types for which two things are true:
+    ///
+    /// - There are two or more different types with the same SQLite representation. E.g.,
+    ///   ValueType::Boolean shares a representation with Integer and Ref.
+    /// - There is no attribute constraint present in the CC.
+    ///
+    /// It's possible at this point for the space of acceptable type tags to not intersect: e.g.,
+    /// for the query
+    ///
+    /// ```edn
+    /// [:find ?x :where
+    ///  [?x ?y true]
+    ///  [?z ?y ?x]]
+    /// ```
+    ///
+    /// where `?y` must simultaneously be a ref-typed attribute and a boolean-typed attribute. This
+    /// function deduces that and calls `self.mark_known_empty`. #293.
+    pub fn expand_type_tags(&mut self) {
```

Comment:
Yes, they're used again, most notably by the projector: for each projected element (e.g., `?v`) it needs to know either its known type or which additional SQL column to project to extract it.